torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -405,49 +402,126 @@ def runtime_wrapper(args: List[Any]):
-):
-    if metadata.is_rng_op_functionalized:
-        assert metadata.num_outputs_rng_offset == 1
-        new_rng_offset = outs[-1]
-        CUDARngStateHelper.set_new_offset(new_rng_offset)
-        if return_new_outs:
-            user_outs = outs[:-1]
-            return user_outs
-        else:
-            return None
-    return outs
+@dataclass
+class FunctionalizedRngRuntimeWrapper(CompilerWrapper):
+    # TODO: I would love to get rid of this argument, but it's
+    # Wrapped pretty tightly around our aot_dispatch_autograd logic.
+    # Specifically, tensors_saved_for_backwards_slice's value is both used for calculating indices
+    # for setting placeholder strides(which is done before runtime, before this wrapper runs)
+    # and for saving tensors for backward (which is done during runtime, after this wrapper runs)
+    # So in aot_dispatch_autograd, this wrapper can't edit the set of outs without making one
+    # of those two indices incorrect.
+    return_new_outs: bool = True
+
+    def post_compile(
+        self,
+        compiled_fn,
+        aot_config: AOTConfig,
+        *,
+        fw_metadata: ViewAndMutationMeta,
+    ):
+        @wraps(compiled_fn)
+        def wrapper(runtime_args: List[Any]):
+            if fw_metadata.is_rng_op_functionalized:
+                # Add the seed and offset to args
+                seed, offset = CUDARngStateHelper.get_torch_state_as_tuple()
+                runtime_args.extend([seed, offset])
+                out = compiled_fn(runtime_args)
+                out = self._functionalized_rng_runtime_epilogue(
+                    fw_metadata,
+                    out,
+                    # TODO: this won't be right for the backward when we convert the call_compiled_backward to use the wrapper
+                    fw_metadata.num_forward_returns,
+                )
+                return out
+            return compiled_fn(runtime_args)
+
+        return wrapper
+
+    # Calling convention: If we are running functionalized RNG, then outs consists
+    # of (user_outs, rng_offset)
+    def _functionalized_rng_runtime_epilogue(
+        self,
+        metadata: ViewAndMutationMeta,
+        outs,
+        offset_index,
+    ):
+        if metadata.is_rng_op_functionalized:
+            assert metadata.num_outputs_rng_offset == 1
+            new_rng_offset = outs[offset_index]
+            CUDARngStateHelper.set_new_offset(new_rng_offset)
+            if self.return_new_outs:
+                user_outs = outs[:offset_index] + outs[offset_index + 1 :]
+                return user_outs
+            else:
+                return outs
+
+        return outs
+
+
+@dataclass
+class FakifiedOutWrapper(CompilerWrapper):
+    fakified_out: Optional[Any] = None
+
+    def post_compile(
+        self,
+        compiled_fn,
+        aot_config: AOTConfig,
+        *,
+        fw_metadata: ViewAndMutationMeta,
+    ):
+        fakified_out = self.fakified_out
```

Comment:
Hmm - I think you mentioned this over chat but just to restate: I guess with warm caching, the idea is that we would include the `fakify_first_call` config in the cache key, and (conditionally) need to include the corresponding `fakified_out` in the cache value so we can return it here.

I think that would work. Curious if @eellison has any thoughts:

(1) warm caching definitely makes the fakify_first_call logic more complicated. At first I was thinking that we could just skip the cache all-together when that config is used, but that wouldn't be easier - by default, that would require us to run all of AOTAutograd's tracing logic whenever the config is used, which kinda defeats the purpose of the warm cache

(2) Do you think the `fakify_first_call` logic is sufficiently useful to be general infra? Having a way to get out the metadata from AOTAutograd's output without using real inputs does seem useful. Then again, it's only used by ddp_optimizer today (which... we can maybe kill now that compiled autograd ddp works? Then again, we can't turn compiled autograd on by default, so anybody not using compiled autograd needs to use the ddp_optimizer...)

Reply:
Is the end goal of fakify_first_call to make sure that the strides of the output of a compiled module are correct? If so, wouldn't a cached, non fake result always have the correct strides given the same input module(i.e. the same cache key)? 

If so, on warm start, we would be able to ignore fakify_first_call, because on a cache hit we'd know the strides were already correct?

FWIW I'm having a bit of trouble reading the notes related to fakifying the first call: I see a comment like `See Note: [Fake Modules and AOTAutograd]` but can't find the corresponding note, so maybe I don't have all the context.

Reply:
That note (or lack thereof) already existed when I made changes: https://github.com/pytorch/pytorch/commit/9c9bde515ceb8ad1164b0dfe0b01530f3a0c85ae#diff-a885ebf62b3ac8ad62acdbca474dff7e7a7daa26d28f88278bfd8514d3168c4cL509. 

But this explainer should suffice: https://github.com/pytorch/pytorch/blob/e28d9947a13493f8f71249fb40cf608e2c5034b5/torch/_dynamo/backends/distributed.py#L318

Reply:
> If so, wouldn't a cached, non fake result always have the correct strides given the same input module

Yes - but I don't see how this would work with the FakeMode first pass we do in compilation. Inductor outputs do not compose with TorchdispatchMode because we execute a fast path to do allocations. 

See, this first loop here that compiles with fake tensor https://github.com/pytorch/pytorch/blob/e28d9947a13493f8f71249fb40cf608e2c5034b5/torch/_dynamo/backends/distributed.py#L363-L365

Reply:
So I guess in my head, at a high level, it would be something like:
- On cache miss: we run the fake mode first pass, but don't save that to cache. We then do a second pass with real values and correct strides, save that result to cache. 
- On a cache hit: we can skip that first pass now and immediately return the cache hit, since we know the strides are correct.